File: tmp/inertia_fixer.py
"""
Fix broken (non-physical) inertias in SDFormat/URDF,
by writing a revised file.
"""
from __future__ import annotations

import abc
import argparse
from dataclasses import dataclass
from typing import Callable
import xml.parsers.expat as expat

import numpy as np

from pydrake.geometry import (Role, SceneGraph)
from pydrake.multibody.plant import MultibodyPlant
from pydrake.multibody.parsing import Parser
from pydrake.multibody.tree import (BodyIndex, CalcSpatialInertia, RigidBody,
                                    SpatialInertia, UnitInertia)


@dataclass
class SourceLocation:
    """The location of an XML syntax item as reported by expat parsing."""

    index: int
    "0-based byte index from start of input stream."

    line: int
    "1-based line number from start of input stream."


@dataclass
class ElementFacts:
    """Facts about an XML element as collecting by expat parsing."""

    name: str
    "The name of the XML element."

    attributes: dict
    "The attributes of the XML element."

    depth: int
    "The nesting depth of the XML element; 0 for the root element."

    start: SourceLocation
    "The location of the start of the XML element text."

    end: SourceLocation = None
    """The location of the end of the XML element text; may require
    adjustments, depending on the input syntax used.
    """

    parent: ElementFacts = None
    "Facts for the parent of this element (optional)."

    serial_number: int = None
    """An opinionated index of this element within the sequence of similar
    elements; see XmlInertiaMapper's implementation for details."""


class FormatDriver(object, metaclass=abc.ABCMeta):

    """This interface encapsulates the differences between URDF and SDFormat
    processing, for purposes of the implementation of XmlInertiaMapper.
    """

    @abc.abstractmethod
    def is_model_element(self, name: str) -> bool:
        """Return True if the element directly encloses link elements."""
        raise NotImplementedError

    @abc.abstractmethod
    def is_element_ignored(self, name: str, attributes: dict) -> bool:
        """Return True if the element is ignored by the Drake parser."""
        raise NotImplementedError

    # TODO: support some per-format element indexing.

    @abc.abstractmethod
    def associate_plant_models(self, models: list, links: list,
                               inertials: list, plant: MultibodyPlant) -> dict:
        """Return a mapping of body_index to inertial ElementFacts."""
        raise NotImplementedError

    # TODO: adjust interface to allow passing inertia pose.
    # Or, maybe split this into mass/pose/inertia APIs?
    @abc.abstractmethod
    def format_inertia(self, mass: float, moments: list, products: list,
                       indenter: Callable[[int], str]) -> str:
        """Return a formatted XML string for the repaired inertial tag.

        The provided products and moments of inertia should be expressed in the
        body frame.

        The indenter function takes an indentation level relative to the
        indentation depth of the element to be replaced, and returns a string
        of white space necessary to reach that indentation level.

        Args:
            mass: the mass of the body (kg).
            moments: the principle moments of inertia -- ixx, iyy, izz.
            products: the products of inertia -- ixy, ixz, iyz.
            indenter: the indenter function (see above).

        """
        raise NotImplementedError


class UrdfDriver(FormatDriver):
    """Format driver for URDF files."""

    def is_model_element(self, name: str) -> bool:
        return name == "robot"

    def is_element_ignored(self, name: str, attributes: dict) -> bool:
        # TODO(rpoyner_tri): The 'drake_ignore' attribute is regrettable legacy
        # cruft that should be removed when the associated URDF parser cruft is
        # removed.
        return ((attributes.get("drake_ignore") == "true")
                or (attributes.get("name") == "world"))

    def associate_plant_models(self, models: list, links: list,
                               inertials: list, plant: MultibodyPlant) -> dict:
        # This assertion is weakened for files that use 'drake_ignore'.
        assert len(links) >= plant.num_bodies() - 1, (
            links, plant.num_bodies())
        mapping = {}
        for inertial in inertials:
            link = inertial.parent
            if link.serial_number is None:
                continue
            k = 1 + link.serial_number
            assert k < plant.num_bodies()
            mapping[BodyIndex(k)] = inertial
            # TODO assert more sanity.
        return mapping

    def format_inertia(self, mass: float, moments: list, products: list,
                       indenter: Callable[[int], str]) -> str:
        d = indenter
        return f"""\
<inertial>
{d(1)}<mass value="{mass}"/>
{d(1)}<origin xyz="0 0 0" rpy="0 0 0"/>
{d(1)}<inertia ixx="{moments[0]}" ixy="{products[0]}" ixz="{products[1]}"\
 iyy="{moments[1]}" iyz="{products[2]}" izz="{moments[2]}"/>
{d(0)}</inertial>"""


class SdformatDriver(FormatDriver):
    """Format driver for SDFormat files."""

    def is_model_element(self, name: str) -> bool:
        return name == "model"

    def is_element_ignored(self, name: str, attributes: dict) -> bool:
        return False

    def associate_plant_models(self, models: list, links: list,
                               inertials: list, plant: MultibodyPlant) -> dict:
        # Because SDFormat has both nested models and inclusions, we will have
        # to rummage around in the plant, finding body indices by using name
        # strings.
        mapping = {}
        for inertial in inertials:
            link = inertial.parent
            model = link.parent
            enclosing_models = [model]
            while model.parent:
                model = model.parent
                enclosing_models.append(model)
            enclosing_models.reverse()
            model_name = '::'.join(
                [m.attributes["name"] for m in enclosing_models])
            model_instance = plant.GetModelInstanceByName(model_name)
            body = plant.GetBodyByName(link.attributes["name"], model_instance)
            mapping[body.index()] = inertial
            # TODO assert more sanity.
        return mapping

    def format_inertia(self, mass: float, moments: list, products: list,
                       indenter: Callable[[int], str]) -> str:
        d = indenter
        return f"""\
<inertial>
{d(1)}<pose>0 0 0 0 0 0</pose>
{d(1)}<mass>{mass}</mass>
{d(1)}<inertia>
{d(2)}<ixx>{moments[0]}</ixx>
{d(2)}<ixy>{products[0]}</ixy>
{d(2)}<ixz>{products[1]}</ixz>
{d(2)}<iyy>{moments[1]}</iyy>
{d(2)}<iyz>{products[2]}</iyz>
{d(2)}<izz>{moments[2]}</izz>
{d(1)}</inertia>
{d(0)}</inertial>"""


class XmlInertiaMapper:
    """Handles the parsing, indexing, and output generation details for editing
    of inertias in some formats of XML robot model files.
    """

    def __init__(self, input_text: str):
        self._input_text = input_text

        # Build a line array for later use.  Add an unused "line" to simulate
        # 1-based indexing.
        self._input_lines = [""] + self._input_text.split('\n')

        # Configure the parser.
        self._parser = expat.ParserCreate()
        self._parser.StartElementHandler = self._start_element
        self._parser.EndElementHandler = self._end_element

        # Declare some state to remember during parsing.
        self._depth = 0
        self._models = []
        self._model_stack = []
        self._links = []
        self._ignored_links = 0
        self._inertials = []
        self._format_driver = None

        # Eventually build a mapping from body_index to inertial_facts.
        self._mapping = {}

    def _make_location(self) -> SourceLocation:
        return SourceLocation(self._parser.CurrentByteIndex,
                              self._parser.CurrentLineNumber)

    def _make_element_facts(self, name: str, attributes: dict) -> ElementFacts:
        return ElementFacts(name, attributes,
                            self._depth, self._make_location())

    def _start_element(self, name: str, attributes: dict):
        if not self._format_driver and self._depth == 0:
            if name == "robot":
                self._format_driver = UrdfDriver()
            elif name == "sdf":
                self._format_driver = SdformatDriver()
            else:
                raise RuntimeError("unknown file format!")

        if self._format_driver.is_model_element(name):
            element = self._make_element_facts(name, attributes)
            if self._model_stack:
                element.parent = self._model_stack[-1]
            self._models.append(element)
            self._model_stack.append(element)

        if name == "link":
            element = self._make_element_facts(name, attributes)
            element.parent = self._model_stack[-1]
            element.serial_number = len(self._links) - self._ignored_links

            if self._format_driver.is_element_ignored(name, attributes):
                self._ignored_links += 1
                element.serial_number = None

            self._links.append(element)

        if name == "inertial":
            element = self._make_element_facts(name, attributes)
            element.parent = self._links[-1]
            self._inertials.append(element)

        self._depth += 1

    def _end_element(self, name: str):
        self._depth -= 1

        if self._format_driver.is_model_element(name):
            self._model_stack.pop()

        if name == "inertial":
            self._inertials[-1].end = self._make_location()

    def parse(self):
        """Execute the parsing of the XML text."""
        self._parser.Parse(self._input_text)

    def associate_plant_models(self, plant: MultibodyPlant):
        """Match body indices to inertial elements in the input text."""
        assert self._format_driver is not None
        self._mapping = self._format_driver.associate_plant_models(
            self._models, self._links, self._inertials, plant)

    def mapping(self) -> dict:
        """Return a mapping from body indices to inertial element facts."""
        return self._mapping

    def _adjusted_element_end_index(self, raw_end_index: int) -> int:
        input_text = self._input_text
        if input_text[raw_end_index] == '<':
            # Typical case:
            # `<inertial>...</inertial>`.
            #               ^                # Raw index points here.
            #                          ^     # Adjusted index points here.
            return raw_end_index + input_text[raw_end_index:].find('>') + 1
        else:
            # `<inertial/>` corner case.
            #             ^     # Raw index points here; good to go
            return raw_end_index

    def build_output(self, new_inertias_mapping: dict) -> str:
        input_text = self._input_text
        output = ""
        index = 0
        for body_index, new_inertia in sorted(new_inertias_mapping.items()):
            facts = self._mapping[body_index]
            output += input_text[index:facts.start.index]
            output += self._format_inertia(body_index, new_inertia)
            index = self._adjusted_element_end_index(facts.end.index)
        output += input_text[index:]
        return output

    def _format_inertia(self, body_index: BodyIndex,
                        spatial_inertia: SpatialInertia) -> str:
        inertial_facts = self._mapping[body_index]

        # Compute indentation based on the first line of the target element.
        line = self._input_lines[inertial_facts.start.line]
        spaces = len(line) - len(line.lstrip())
        indentation = spaces // inertial_facts.depth

        # Extract the mass properties.
        mass = spatial_inertia.get_mass()
        rot = spatial_inertia.CalcRotationalInertia()
        mom = rot.get_moments()
        prod = rot.get_products()

        # Build the indenter.
        depth = inertial_facts.depth
        indent = " " * indentation

        def d(more_depth):
            return indent * (depth + more_depth)

        return self._format_driver.format_inertia(mass, mom, prod, d)


class InertiaProcessor:
    """Handles selection, repair, and replacement of inertial properties,
    in model files pre-processed by drake parsing and XmlInertiaMapper.
    """

    def __init__(self, args: argparse.Namespace, plant: MultibodyPlant,
                 scene_graph: SceneGraph, mapper: XmlInertiaMapper):
        self._args = args
        self._plant = plant
        self._scene_graph = scene_graph
        self._mapper = mapper

    def _maybe_fix_inertia(
            self, body_index: BodyIndex) -> SpatialInertia | None:
        assert int(body_index) < self._plant.num_bodies(), (
            body_index, self._plant.num_bodies())
        body = self._plant.get_body(body_index)
        if not isinstance(body, RigidBody):
            # Only rigid bodies have constant inertia, for which model file
            # fixups make sense.
            return
        maybe_frame_id = self._plant.GetBodyFrameIdIfExists(body_index)
        if not maybe_frame_id:
            # No geometry to fix inertia from.
            return
        old_inertia = body.default_spatial_inertia()
        if self._args.invalid_only and old_inertia.IsPhysicallyValid():
            # Skip valid inertias by user preference.
            return
        inspector = self._scene_graph.model_inspector()
        geoms = inspector.GetGeometries(maybe_frame_id, Role.kProximity)
        if not geoms:
            # No geometry to fix inertia from.
            return

        # Collect some density==1 inertias for all geometries.
        M_BBo_B_one = SpatialInertia(0, np.zeros(3), UnitInertia(0, 0, 0))
        for geom in geoms:
            M_GG_G_one = CalcSpatialInertia(inspector.GetShape(geom), 1.0)
            X_BG = inspector.GetPoseInFrame(geom)
            M_GBo_B_one = M_GG_G_one.ReExpress(
                X_BG.rotation()).Shift(-X_BG.translation())
            M_BBo_B_one += M_GBo_B_one

            # Rebuild the final inertia using the mass found in the input.
            mass = old_inertia.get_mass()
            M_BBo_B = SpatialInertia(mass, M_BBo_B_one.get_com(),
                                     M_BBo_B_one.get_unit_inertia())

        # TODO: shift the summed-up inertia back to Bcm, and return that.

        return M_BBo_B

    def process(self) -> str:
        """Return a new model file text, with selected inertial properties
        replaced by new ones calculated from supplied geometry.
        """
        mapping = self._mapper.mapping()
        new_inertias_mapping = {}
        for body_index in sorted(mapping.keys()):
            maybe_inertia = self._maybe_fix_inertia(body_index)
            if maybe_inertia:
                new_inertias_mapping[body_index] = maybe_inertia
        return self._mapper.build_output(new_inertias_mapping)


def main():
    # Parse arguments.
    args_parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter)
    args_parser.add_argument(
        "input_file", type=str,
        help="Filesystem path to an SDFormat or URDF file.")
    args_parser.add_argument(
        "output_file", type=str, nargs='?',
        help="[Optional] Filesystem path to write output with repaired"
        " inertias. If missing, output will go to stdout.")
    args_parser.add_argument(
        "--invalid_only", action="store_true",
        help="only fix physically invalid inertias.")
    args_parser.add_argument(
        "--in_place", action="store_true",
        help="modify the input file in-place. Any output_file argument"
        " will be ignored.")
    args = args_parser.parse_args()

    # Parse with drake to build mbp and confirm sanity.
    plant = MultibodyPlant(time_step=0.0)
    scene_graph = SceneGraph()
    plant.RegisterAsSourceForSceneGraph(scene_graph)

    parser = Parser(plant)
    parser.package_map().PopulateFromRosPackagePath()

    # Read from the disk file here, to get more lenient processing of URIs,
    # better error messages, etc.
    parser.AddModels(args.input_file)

    # Slurp input file for indexing and editing.
    with open(args.input_file) as fo:
        input_text = fo.read()

    # Parse with expat to build index.
    mapper = XmlInertiaMapper(input_text)
    mapper.parse()
    mapper.associate_plant_models(plant)

    # Fix indicated inertias.
    processor = InertiaProcessor(args, plant, scene_graph, mapper)
    output_text = processor.process()

    # Write output.
    if args.output_file:
        output_file = args.output_file
    else:
        output_file = "/dev/stdout"
    if args.in_place:
        output_file = args.input_file
    with open(output_file, 'w') as fo:
        fo.write(output_text)


if __name__ == '__main__':
    main()

File: tmp/inertia_fixer.cc
#include <fstream>
#include <functional>

#include <drake_vendor/tinyxml2.h>
#include <gflags/gflags.h>

#include "drake/multibody/parsing/parser.h"
#include "drake/multibody/tree/geometry_spatial_inertia.h"

DEFINE_bool(invalid_only, false,
            "if true, only fix physically invalid inertias");

namespace drake {
namespace multibody {
namespace {

using geometry::Role;
using geometry::SceneGraph;
using tinyxml2::XMLNode;
using tinyxml2::XMLDocument;
using tinyxml2::XMLElement;
using tinyxml2::XMLPrinter;

// XXX Does tinyxml2 quote-swapping ('' to "") break any strings?

// Nope: it does entity substitution. So things may look worse in source, but
// information will be preserved.
/*
For Input:
  <material name='silver-"ish"'>

We get Output:
  <material name="silver-&quot;ish&quot;">
*/

// We jump through hoops here, just to get 2-space indented output.
class XmlPrinter final : public XMLPrinter {
 public:
  XmlPrinter() : XMLPrinter() {}
  void PrintSpace(int depth) final {
    for (int i = 0; i < depth; ++i) {
      Write("  ");
    }
  }
};

// Create a child element if it does not already exist.
XMLElement* EnsureChildElement(XMLElement* el, const char* child) {
  XMLElement* kid = el->FirstChildElement(child);
  if (!kid) {
    kid = el->InsertNewChildElement(child);
  }
  return kid;
}

// Format a double in shortest round-trip representation.
std::string roundtrip(double x) {
  return fmt::format("{}", x);
}

// Recursively find all of the descendant elements of 'el` named "link", and
// accumulate them in an externally supplied vector.
void FindLinks(XMLElement* el, std::vector<XMLElement*>* links) {
  if (std::string(el->Name()) == "link") {
    links->push_back(el);
    return;
  }
  for (XMLElement* kid = el->FirstChildElement();
       kid;
       kid = kid->NextSiblingElement()) {
    FindLinks(kid, links);
  }
}


class XmlInertiaMapper {
 public:
  using Mapping = std::unordered_map<BodyIndex, XMLElement*>;
  virtual ~XmlInertiaMapper() {}
  virtual const Mapping& mapping() const = 0;
  virtual void UpdateInertia(XMLElement*, const SpatialInertia<double>&) = 0;
};

class UrdfInertiaMapper final : public XmlInertiaMapper {
 public:
  UrdfInertiaMapper(
      const MultibodyPlant<double>& plant,
      const std::vector<ModelInstanceIndex>& models,
      XMLDocument* doc)
      : plant_(plant) {
    DRAKE_ASSERT(models.size() == 1);
    std::vector<XMLElement*> links;
    FindLinks(doc->RootElement(), &links);
    DRAKE_ASSERT(drake::ssize(links) == plant_.num_bodies() - 1);
    for (int k = 1; k < plant_.num_bodies(); ++k) {
      const BodyIndex body_index{k};
      mapping_[body_index] = links[k - 1];
      if (kDrakeAssertIsArmed) {
        const auto& body_name = plant_.get_body(body_index).name();
        const XMLElement* link = links[k - 1];
        const char* name = link->Attribute("name");
        DRAKE_ASSERT(name != nullptr);
        DRAKE_ASSERT(std::string(name) == body_name);
      }
    }
  }

  const Mapping& mapping() const final { return mapping_; }

  void UpdateInertia(XMLElement* el,
                     const SpatialInertia<double>& inertia) final {
    XMLElement* inertial = EnsureChildElement(el, "inertial");
    XMLElement* origin = EnsureChildElement(inertial, "origin");
    origin->SetAttribute("xyz", "0 0 0");
    origin->SetAttribute("rpy", "0 0 0");
    XMLElement* mass = EnsureChildElement(inertial, "mass");
    mass->SetAttribute("value", roundtrip(inertia.get_mass()).c_str());
    XMLElement* coeffs = EnsureChildElement(inertial, "inertia");
    const auto rot = inertia.CalcRotationalInertia();
    const auto mom = rot.get_moments();
    const auto prod = rot.get_products();
    coeffs->SetAttribute("ixx", roundtrip(mom(0)).c_str());
    coeffs->SetAttribute("iyy", roundtrip(mom(1)).c_str());
    coeffs->SetAttribute("izz", roundtrip(mom(2)).c_str());
    coeffs->SetAttribute("ixy", roundtrip(prod(0)).c_str());
    coeffs->SetAttribute("ixz", roundtrip(prod(1)).c_str());
    coeffs->SetAttribute("iyz", roundtrip(prod(2)).c_str());
  }

 private:
  const MultibodyPlant<double>& plant_;
  Mapping mapping_;
};


class SdfInertiaMapper final : public XmlInertiaMapper {
 public:
  SdfInertiaMapper(
      const MultibodyPlant<double>& plant,
      const std::vector<ModelInstanceIndex>& models,
      XMLDocument* doc)
      : plant_(plant) {
    DRAKE_ASSERT(models.size() >= 1);
    std::vector<XMLElement*> links;
    FindLinks(doc->RootElement(), &links);
    // XXX This assumption is likely violated by files that use include tags.
    DRAKE_ASSERT(drake::ssize(links) == plant_.num_bodies() - 1);
    for (int k = 1; k < plant_.num_bodies(); ++k) {
      const BodyIndex body_index{k};
      mapping_[body_index] = links[k - 1];
      if (kDrakeAssertIsArmed) {
        const auto& body_name = plant_.get_body(body_index).name();
        const XMLElement* link = links[k - 1];
        const char* name = link->Attribute("name");
        DRAKE_ASSERT(name != nullptr);
        DRAKE_ASSERT(std::string(name) == body_name);
      }
    }
  }

  const Mapping& mapping() const final { return mapping_; }

  void UpdateInertia(XMLElement* el,
                     const SpatialInertia<double>& inertia) final {
    // Fill out the inertial properties.
    XMLElement* inertial = EnsureChildElement(el, "inertial");
    XMLElement* pose = EnsureChildElement(inertial, "pose");
    pose->SetText("0 0 0 0 0 0");
    auto write_child = [&](XMLElement* parent, const char* name, double value) {
      XMLElement* kid = EnsureChildElement(parent, name);
      kid->SetText(roundtrip(value).c_str());
    };
    write_child(inertial, "mass", inertia.get_mass());
    XMLElement* coeffs = EnsureChildElement(inertial, "inertia");
    const auto rot = inertia.CalcRotationalInertia();
    const auto mom = rot.get_moments();
    const auto prod = rot.get_products();
    write_child(coeffs, "ixx", mom(0));
    write_child(coeffs, "iyy", mom(1));
    write_child(coeffs, "izz", mom(2));
    write_child(coeffs, "ixy", prod(0));
    write_child(coeffs, "ixz", prod(1));
    write_child(coeffs, "iyz", prod(2));
  }

 private:
  const MultibodyPlant<double>& plant_;
  Mapping mapping_;
};


class InertiaProcessor {
 public:
  InertiaProcessor(
      const MultibodyPlant<double>& plant,
      const SceneGraph<double>& scene_graph,
      const std::vector<ModelInstanceIndex>& models,
      XMLDocument* doc)
      : plant_(plant), scene_graph_(scene_graph) {
    XMLElement* root = doc->RootElement();
    std::string root_name(root->Name());
    if (root_name == "sdf") {
      mapper_ = std::make_unique<SdfInertiaMapper>(plant, models, doc);
    } else if (root_name == "robot") {
      mapper_ = std::make_unique<UrdfInertiaMapper>(plant, models, doc);
    } else {
      drake::log()->error("Unknown file type: root element was {}", root_name);
      ::exit(EXIT_FAILURE);
    }
  }

  void Process() {
    for (const auto& pair : mapper_->mapping()) {
      const auto& [body_index, el] = pair;
      const auto maybe_inertia = MaybeFixBodyInertia(body_index);
      if (maybe_inertia.has_value()) {
        mapper_->UpdateInertia(el, *maybe_inertia);
      }
    }
  }

 private:
  std::optional<SpatialInertia<double>> MaybeFixBodyInertia(
      BodyIndex body_index) {
    const auto& body = plant_.get_body(body_index);
    const auto* rigid_body = dynamic_cast<const RigidBody<double>*>(&body);
    if (!rigid_body) {
      // Only rigid bodies have constant inertia, for which model file fixups
      // make sense.
      return {};
    }
    const auto maybe_frame_id = plant_.GetBodyFrameIdIfExists(body_index);
    if (!maybe_frame_id.has_value()) {
      // No geometry to fix inertia from.
      return {};
    }
    const auto& old_inertia = rigid_body->default_spatial_inertia();
    if (FLAGS_invalid_only && old_inertia.IsPhysicallyValid()) {
      // Skip valid inertias by user preference.
      return {};
    }
    const auto& inspector = scene_graph_.model_inspector();
    const auto geoms = inspector.GetGeometries(*maybe_frame_id,
                                               Role::kProximity);
    if (geoms.empty()) {
      // No geometry to fix inertia from.
      return {};
    }
    SpatialInertia<double> M_BBo_B(0, {0, 0, 0}, {0, 0, 0});
    // XXX mass calculations are all wrong for the case of multiple geometries.
    const double mass = old_inertia.get_mass();
    for (const auto& geom : geoms) {
      const auto M_GG_G_one = CalcSpatialInertia(inspector.GetShape(geom), 1.0);
      SpatialInertia<double> M_GG_G(mass, M_GG_G_one.get_com(),
                                    M_GG_G_one.get_unit_inertia());
      const auto& X_BG = inspector.GetPoseInFrame(geom);
      const auto M_GBo_B = M_GG_G.ReExpress(X_BG.rotation())
                           .Shift(-X_BG.translation());
      M_BBo_B += M_GBo_B;
    }
    return M_BBo_B;
  }

  const MultibodyPlant<double>& plant_;
  const SceneGraph<double>& scene_graph_;
  std::unique_ptr<XmlInertiaMapper> mapper_;
};

int do_main(int argc, char* argv[]) {
  gflags::SetUsageMessage("[INPUT-FILE] [OUTPUT-FILE]\n"
                          "Rewrite URDF/SDFormat with fixed-up inertias");
  gflags::ParseCommandLineFlags(&argc, &argv, true);
  if (argc < 2) {
    drake::log()->error("missing input filename");
    return 1;
  }
  const char* outfile = (argc > 2) ? argv[2] : "/dev/stdout";

  XMLDocument xml_doc;
  xml_doc.LoadFile(argv[1]);
  if (xml_doc.ErrorID()) {
    drake::log()->error("Failed to parse XML file: {}",
                        xml_doc.ErrorName());
    ::exit(EXIT_FAILURE);
  }

  MultibodyPlant<double> plant{0.0};
  SceneGraph<double> scene_graph;
  plant.RegisterAsSourceForSceneGraph(&scene_graph);

  Parser parser{&plant};
  parser.package_map().PopulateFromRosPackagePath();

  // Chances are good that parsing may end in an error (implemented as throw).
  // For purposes of this tool, we should make its message usable. To do that,
  // catch it and display it via logging. Yes, this violates the coding
  // standard; in this case it makes the tool significantly more useful.
  std::vector<ModelInstanceIndex> models;
  try {
    models = parser.AddModels(argv[1]);
  } catch (const std::exception& e) {
    drake::log()->error(e.what());
    ::exit(EXIT_FAILURE);
  }

  InertiaProcessor processor(plant, scene_graph, models, &xml_doc);
  processor.Process();

  // Use our custom XmlPrinter to "print to memory", then dump that to file.
  XmlPrinter printer;
  xml_doc.Print(&printer);
  std::ofstream out(outfile);
  out << printer.CStr();

  return EXIT_SUCCESS;
}

}  // namespace
}  // namespace multibody
}  // namespace drake

int main(int argc, char* argv[]) {
  return drake::multibody::do_main(argc, argv);
}
